# Analysing the Bike Station Data for Lyon from JCDecaux API

#### URL used to fetch data: 
https://api.jcdecaux.com/vls/v1/stations?contract=Lyon&apiKey=b1c702f23769a30bf7f040b18ce456da6301cd4d

<div style="text-align: right;">
    <h4>Sumit Trivedi -- K00302092</h4>
    <h4>Moutrisha Ghosh -- K00302103</h4>
</div>

## Introduction

plan for analysing the data 
Include the plan at the top / beginning of the notebook.

## Project Plan and Analysis Goals

### Plans:

### Analysis Goals:

1. Combine the data from all three days

    
3. Before doing any analysis change the time format from miliseconds to proper format
    

## Data Preprocessing

In [537]:
import pandas as pd # Importing the pandas library 
import numpy as np # Importing the numpy library 
import datetime # Importing the datetime module to work with date and time objects

We have collected the data for a span of approximately 12 hours for 3 days  (21st Dec, 22nd Dec, 23rd Dec) and stored in 3 different csv files for each day. We have loaded and concatenated the files into a single dataframe below:

In [540]:
sat = pd.read_csv(r'Saturday_bike_data.csv') # Reading the CSV file with data retrived on Saturday
sun = pd.read_csv(r'Sunday_bike_data.csv') # Reading the CSV file with data retrived on Sunday
mon = pd.read_csv(r'Monday_bike_data.csv') # Reading the CSV file with data retrived on Monday

combined_bike_df = pd.concat([sat, sun, mon], ignore_index = True) # Combining the data from the three days into a single DataFrame.

In [541]:
combined_bike_df.head() # Displaying the first five rows of the combined DataFrame to preview the data

number contract_name                           name  \
0    2010          lyon      2010 - CONFLUENCE / DARSE   
1    5015          lyon               5015 - FULCHIRON   
2   32001          lyon        32001 - COUZON - CENTRE   
3    6004          lyon                    6004 - FOCH   
4    7035          lyon  7035 - MARSEILLE / UNIVERSITÉ   

                                         address  \
0  ANGLE ALLEE ANDRE MURE ET QUAI ANTOINE RIBOUD   
1           Devant le n°41 rue de la Quarantaine   
2                                            NaN   
3                              Angle rue Vendôme   
4                         Angle rue de Marseille   

                              position  banking  bonus  bike_stands  \
0  {'lat': 45.743317, 'lng': 4.815747}     True  False           22   
1   {'lat': 45.75197, 'lng': 4.821662}     True  False           19   
2  {'lat': 45.846034, 'lng': 4.832409}     True  False           17   
3  {'lat': 45.768896, 'lng': 4.844845}     True  False           20   
4   {'lat': 45.750945, 'lng': 4.83927}     True  False           40   

   available_bike_stands  available_bikes status    last_update  
0                     21                1   OPEN  1734775168000  
1                      6               11   OPEN  1734775441000  
2                      3               14   OPEN  1734775201000  
3                      9               10   OPEN  1734775462000  
4                     34                4   OPEN  1734775244000

In [542]:
combined_bike_df.tail() # Displaying the last five rows of the combined DataFrame to inspect the data at the end

number contract_name                                         name  \
207995    8002          lyon                8002- PLACE AMBROISE COURTOIS   
207996   10007          lyon                   10007 - PIATON / CONDORCET   
207997    9033          lyon                         9033 - GORGE DE LOUP   
207998    5001          lyon  5001 - PLACE VARILLON (FUNICULAIRE ST JUST)   
207999    3018          lyon                     3018 - CRÉQUI / VOLTAIRE   

                            address                             position  \
207995  Face au cours Albert Thomas  {'lat': 45.745296, 'lng': 4.871604}   
207996  Av Piaton / angle Condorcet  {'lat': 45.774204, 'lng': 4.867512}   
207997           49 RUE SGT BERTHET  {'lat': 45.767484, 'lng': 4.805626}   
207998  Devant le n°63 rue de Trion  {'lat': 45.757319, 'lng': 4.815064}   
207999          Angle rue Paul Bert  {'lat': 45.756596, 'lng': 4.848442}   

        banking  bonus  bike_stands  available_bike_stands  available_bikes  \
207995     True  False           20                     10                9   
207996    False  False           20                     15                5   
207997     True  False           40                     39                1   
207998     True   True           23                     23                0   
207999     True  False           30                     11               19   

       status    last_update  
207995   OPEN  1734988553000  
207996   OPEN  1734988304000  
207997   OPEN  1734988820000  
207998   OPEN  1734988747000  
207999   OPEN  1734988643000

In [546]:
type(combined_bike_df) # Checking the type of the combined_bike_df variable to confirm it is a pandas DataFrame

pandas.core.frame.DataFrame

In [548]:
combined_bike_df.info() # Displaying the summary information of the combined DataFrame to help in understanding the structure of the data

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 208000 entries, 0 to 207999
Data columns (total 12 columns):
 #   Column                 Non-Null Count   Dtype 
---  ------                 --------------   ----- 
 0   number                 208000 non-null  int64 
 1   contract_name          208000 non-null  object
 2   name                   208000 non-null  object
 3   address                166000 non-null  object
 4   position               208000 non-null  object
 5   banking                208000 non-null  bool  
 6   bonus                  208000 non-null  bool  
 7   bike_stands            208000 non-null  int64 
 8   available_bike_stands  208000 non-null  int64 
 9   available_bikes        208000 non-null  int64 
 10  status                 208000 non-null  object
 11  last_update            208000 non-null  int64 
dtypes: bool(2), int64(5), object(5)
memory usage: 16.3+ MB


Checking the Summary Statistics of the dataframe before we proceed with the Data Preprocessing and Analysis:

In [551]:
combined_bike_df.describe() # Generating summary statistics for the numerical columns in the combined DataFrame

number    bike_stands  available_bike_stands  available_bikes  \
count  208000.000000  208000.000000          208000.000000    208000.000000   
mean     7469.192308      21.074519              11.486135         8.883601   
std      5448.301986       7.344346               7.771656         7.985367   
min       202.000000       1.000000               0.000000         0.000000   
25%      3093.000000      16.000000               5.000000         2.000000   
50%      7019.500000      20.000000              11.000000         7.000000   
75%     10019.500000      24.000000              16.000000        14.000000   
max     34002.000000      55.000000              58.000000        40.000000   

        last_update  
count  2.080000e+05  
mean   1.734877e+12  
std    1.728356e+08  
min    1.732269e+12  
25%    1.734812e+12  
50%    1.734891e+12  
75%    1.734956e+12  
max    1.734989e+12

In [553]:
data = combined_bike_df.copy() #creating a copy that we will be working on while the original data remains same

In [555]:
data.shape # Checking the shape of the DataFrame to see the number of rows and columns

(208000, 12)

Identifying the duplicate rows in the DataFrame and processing it accordingly:

In [558]:
duplicates = data[data.duplicated()] # Identifying duplicate rows 
duplicates.sort_values(by = 'last_update') # Sorting the duplicate rows by the 'last_update' column

number contract_name                          name  \
138256   10049          lyon  10049 - BORNE TEST A. THOMAS   
41328    10049          lyon  10049 - BORNE TEST A. THOMAS   
145744   10049          lyon  10049 - BORNE TEST A. THOMAS   
185264   10049          lyon  10049 - BORNE TEST A. THOMAS   
89584    10049          lyon  10049 - BORNE TEST A. THOMAS   
...        ...           ...                           ...   
207949   10019          lyon       10019 - 4 AOUT / RACINE   
207747    6025          lyon       6025 - PLACE D'HELVÉTIE   
207648    3103          lyon           3103 - PLACE DANTON   
207606    4023          lyon           4023 - PONT MASARYK   
207850    6016          lyon       6016 - VIABERT / THIERS   

                         address  \
138256  Borne test Albert Thomas   
41328   Borne test Albert Thomas   
145744  Borne test Albert Thomas   
185264  Borne test Albert Thomas   
89584   Borne test Albert Thomas   
...                          ...   
207949                       NaN   
207747                       NaN   
207648            3 PLACE DANTON   
207606    Face au 45 quai Gillet   
207850       Angle avenue Thiers   

                                                 position  banking  bonus  \
138256  {'lat': 45.7481858411998, 'lng': 4.86445471733...    False  False   
41328   {'lat': 45.7481858411998, 'lng': 4.86445471733...    False  False   
145744  {'lat': 45.7481858411998, 'lng': 4.86445471733...    False  False   
185264  {'lat': 45.7481858411998, 'lng': 4.86445471733...    False  False   
89584   {'lat': 45.7481858411998, 'lng': 4.86445471733...    False  False   
...                                                   ...      ...    ...   
207949                {'lat': 45.764489, 'lng': 4.877894}    False  False   
207747                {'lat': 45.773232, 'lng': 4.841949}     True  False   
207648  {'lat': 45.7548502313005, 'lng': 4.85748756866...     True  False   
207606                {'lat': 45.778175, 'lng': 4.810716}     True  False   
207850                {'lat': 45.766032, 'lng': 4.862063}    False  False   

        bike_stands  available_bike_stands  available_bikes status  \
138256            1                      0                0   OPEN   
41328             1                      0                0   OPEN   
145744            1                      0                0   OPEN   
185264            1                      0                0   OPEN   
89584             1                      0                0   OPEN   
...             ...                    ...              ...    ...   
207949           30                      3               26   OPEN   
207747           30                     20                9   OPEN   
207648           18                      0               18   OPEN   
207606           22                      9               12   OPEN   
207850           16                      4               11   OPEN   

          last_update  
138256  1732269059000  
41328   1732269059000  
145744  1732269059000  
185264  1732269059000  
89584   1732269059000  
...               ...  
207949  1734988533000  
207747  1734988533000  
207648  1734988536000  
207606  1734988536000  
207850  1734988537000  

[95608 rows x 12 columns]

In [560]:
data = data.drop_duplicates() # Dropping all duplicate rows from the DataFrame to ensure unique entries only.
data.shape # Checking the updated shape of the DataFrame after removing duplicates

(112392, 12)

In [562]:
data.columns #checking the column names

Index(['number', 'contract_name', 'name', 'address', 'position', 'banking',
       'bonus', 'bike_stands', 'available_bike_stands', 'available_bikes',
       'status', 'last_update'],
      dtype='object')

Checking and Handling Missing Values:

In [565]:
data.isnull().sum() # Checking for missing values in each column of the DataFrame

number                       0
contract_name                0
name                         0
address                  21132
position                     0
banking                      0
bonus                        0
bike_stands                  0
available_bike_stands        0
available_bikes              0
status                       0
last_update                  0
dtype: int64

In [567]:
data = data.fillna(value =  "No Adresss Specified") # Filling missing values (NaN) in the DataFrame with the string "No Address Specified"

In [569]:
data.isnull().sum() # Checking again for missing values in the DataFrame to confirm that all NaN values have been filled

number                   0
contract_name            0
name                     0
address                  0
position                 0
banking                  0
bonus                    0
bike_stands              0
available_bike_stands    0
available_bikes          0
status                   0
last_update              0
dtype: int64

Checking the number of unique values in each column of the DataFrame:

In [572]:
for col in data.columns: 
    print(col,len(data[col].unique())) # Checking the unique values

number 416
contract_name 1
name 416
address 329
position 416
banking 2
bonus 2
bike_stands 36
available_bike_stands 56
available_bikes 41
status 2
last_update 74218


Since we are only working with one City (i.e., Lyon, France), hence we are dropping the column as it's not need for further Analysis:

In [575]:
data =data.drop(['contract_name'], axis=1) # Dropping the 'contract_name' column from the DataFrame as it contains only a single unique value "Lyon"

In [577]:
data.head()  # Displaying the first five rows of the DataFrame to inspect the data after the modifications

number                           name  \
0    2010      2010 - CONFLUENCE / DARSE   
1    5015               5015 - FULCHIRON   
2   32001        32001 - COUZON - CENTRE   
3    6004                    6004 - FOCH   
4    7035  7035 - MARSEILLE / UNIVERSITÉ   

                                         address  \
0  ANGLE ALLEE ANDRE MURE ET QUAI ANTOINE RIBOUD   
1           Devant le n°41 rue de la Quarantaine   
2                           No Adresss Specified   
3                              Angle rue Vendôme   
4                         Angle rue de Marseille   

                              position  banking  bonus  bike_stands  \
0  {'lat': 45.743317, 'lng': 4.815747}     True  False           22   
1   {'lat': 45.75197, 'lng': 4.821662}     True  False           19   
2  {'lat': 45.846034, 'lng': 4.832409}     True  False           17   
3  {'lat': 45.768896, 'lng': 4.844845}     True  False           20   
4   {'lat': 45.750945, 'lng': 4.83927}     True  False           40   

   available_bike_stands  available_bikes status    last_update  
0                     21                1   OPEN  1734775168000  
1                      6               11   OPEN  1734775441000  
2                      3               14   OPEN  1734775201000  
3                      9               10   OPEN  1734775462000  
4                     34                4   OPEN  1734775244000

In [579]:
data.columns # Displaying the list of column names in the DataFrame to confirm the structure of the dataset

Index(['number', 'name', 'address', 'position', 'banking', 'bonus',
       'bike_stands', 'available_bike_stands', 'available_bikes', 'status',
       'last_update'],
      dtype='object')

Counting the occurences of each unique values for the columns that have less unique values:

In [582]:
data.status.value_counts() # Counting the unique values in the 'status' column

status
OPEN      112159
CLOSED       233
Name: count, dtype: int64

In [584]:
data.banking.value_counts() # Counting unique values in the 'banking' column

banking
True     58684
False    53708
Name: count, dtype: int64

In [586]:
data.bonus.value_counts() # Counting unique values in the 'bonus' column

bonus
False    102518
True       9874
Name: count, dtype: int64

Converting the last_update column to datetime format for the ease of analysis:

In [589]:
data['last_update'] = pd.to_datetime(data['last_update'], unit='ms') # Converting the 'last_update' column from milliseconds to datetime format



Generating new attributes for detailed analysis:

In [592]:
data['date'] = data['last_update'].dt.date # Creating a new 'date' column by extracting the date from the 'last_update' column

data['hour'] = data['last_update'].dt.hour # Similarly generating 'hour' column

In [594]:
data['latitude'] = data['position'].str.extract(r"'lat':\s*([\d\.\-]+)") # Generating a new 'latitute' column by extracting the lat from the 'position' column

data['longitude'] = data['position'].str.extract(r"'lng':\s*([\d\.\-]+)") # Similarly generating 'longitude' column

In [595]:
data.columns

Index(['number', 'name', 'address', 'position', 'banking', 'bonus',
       'bike_stands', 'available_bike_stands', 'available_bikes', 'status',
       'last_update', 'date', 'hour', 'latitude', 'longitude'],
      dtype='object')

In [598]:
data.date.value_counts()

date
2024-12-21    38696
2024-12-23    37571
2024-12-22    36121
2024-12-20        1
2024-11-22        1
2024-12-01        1
2024-12-17        1
Name: count, dtype: int64

Looking at the unique dates in our dataset, we can understand that there are 2 outliers as we did not collect data on 2024-11-22 or 2024-12-17. Hence, we will remove these 2 instances from our dataset.

In [601]:
data = data [[data['date'] != '2024-12-17'] or [data['date'] != '2024-11-22']]

KeyError: "None of [Index([(True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, ...)], dtype='object')] are in the [columns]"

## Analysis

### 1. How many unique bike stations are available across Lyon?

In [388]:
stations = data['name'].unique()
pd.DataFrame(stations)

0
0                      2010 - CONFLUENCE / DARSE
1                               5015 - FULCHIRON
2                        32001 - COUZON - CENTRE
3                                    6004 - FOCH
4                  7035 - MARSEILLE / UNIVERSITÉ
..                                           ...
411                8002- PLACE AMBROISE COURTOIS
412                   10007 - PIATON / CONDORCET
413                         9033 - GORGE DE LOUP
414  5001 - PLACE VARILLON (FUNICULAIRE ST JUST)
415                     3018 - CRÉQUI / VOLTAIRE

[416 rows x 1 columns]

According to the above analysis, we can see that there are a total of 416 bike stations across Lyon city in France

### 2. Which are the top 10 bike stations based on bike stand capacity?

In [392]:
df1 = data.groupby(by='name')['bike_stands'].mean().sort_values(ascending=False)
top_10_bike_stations1 = df1.head(10).reset_index()
top_10_bike_stations1

name  bike_stands
0                                10002 - INSA         55.0
1  2018 - HOTEL DIEU / PONT DE LA GUILLOTIERE         50.0
2                                1002 - OPÉRA         41.0
3           3097 - AUGAGNEUR / FOSSE AUX OURS         40.0
4               7035 - MARSEILLE / UNIVERSITÉ         40.0
5               2002 - BELLECOUR / ST EXUPÉRY         40.0
6                              7006 - GERLAND         40.0
7                 7003 - GAMBETTA / GARIBALDI         40.0
8                      20001 - OULLINS - GARE         40.0
9                        9033 - GORGE DE LOUP         40.0

These are the bike stations with the highest number of average bike stands. We will dive deep into this analysis on later in the code.

### 3. Which are the top 10 bike stations that are likely to be underutilized?

In [396]:
df2 = data.groupby(by='name')['available_bike_stands'].mean().sort_values(ascending=False)
top_10_bike_stations2 = df2.head(10).reset_index()
top_10_bike_stations2

name  available_bike_stands
0                         10002 - INSA              45.177083
1                 9033 - GORGE DE LOUP              38.049793
2           8001 - UNIVERSITÉ LYON III              32.180556
3   10112 - PLACE DE LA RECONNAISSANCE              30.139373
4        7035 - MARSEILLE / UNIVERSITÉ              29.785256
5       12004 - VAULX - HÔTEL DE VILLE              29.057778
6       3070 - REBATEL / SAINT-MAXIMIN              29.012397
7       7019 - CROIX BARRET / BOURDEIX              28.373444
8         10102 - DOUA / GASTON BERGER              28.176211
9  10056 - 11 NOVEMBRE / GASTON BERGER              28.053061

These are the bike stations that can be most likely to be underutilized and probably are located in areas with lower demand as they have the highest average number of available bike stands. 

### 4. What is the average number of bikes available at each station across Lyon?

In [400]:
df3 = data.groupby(by='name')['available_bikes'].mean().sort_values(ascending=False)
top_bike_stations3 = df3.reset_index()
top_bike_stations3

name  available_bikes
0                               3094 - VILLETTE        35.077170
1                     3014 - PART-DIEU/POMPIDOU        34.338150
2                     2037 - PLACE DES ARCHIVES        32.731343
3    2018 - HOTEL DIEU / PONT DE LA GUILLOTIERE        32.621118
4                   2008 - PLACE ANTONIN PONCET        31.721127
..                                          ...              ...
411                  5055 - BUYER / APOLLINAIRE         0.049774
412                      0-555 - ATELIER VÉLO'V         0.008772
413                10049 - BORNE TEST A. THOMAS         0.000000
414                   10124 - SALENGRO / YVONNE         0.000000
415                   14005 - BRON - UNIVERSITE         0.000000

[416 rows x 2 columns]

From the above dataframe it can be observed that '3094 - VILLETTE' has the highest average number of bikes available throughout the span of 3 days. On the other hand, '10049 - BORNE TEST A. THOMAS', '10124 - SALENGRO / YVONNE', and '14005 - BRON - UNIVERSITE' are the bike stations where the availability of the bikes is absolutely 0.

### 5. Deep dive into the bike stations having zero available bikes consistently

In [404]:
df4 = data.groupby(by='name')['available_bikes'].sum().sort_values(ascending=False)

top_bike_stations4 = df4.reset_index()

empty_bike_stations = top_bike_stations4[top_bike_stations4['available_bikes'] == 0] 

#Grouping the bike stations based on the total number of available bikes and filtering the bike stations with zero availability.

In [406]:
empty_bike_stations 

name  available_bikes
413     10124 - SALENGRO / YVONNE                0
414  10049 - BORNE TEST A. THOMAS                0
415     14005 - BRON - UNIVERSITE                0

These are the empty bike stations that have zero bikes availability throughtout. 

Now let's look at the data of each of the empty station at a time to understand why the availability at the above stations could be zero:

In [409]:
data[data["name"] == "10124 - SALENGRO / YVONNE"]

number                       name  \
140   10124  10124 - SALENGRO / YVONNE   

                                          address  \
140  1 rue Georges Courteline\n69100 Villeurbanne   

                                              position  banking  bonus  \
140  {'lat': 45.7772936957677, 'lng': 4.87675865740...    False  False   

     bike_stands  available_bike_stands  available_bikes  status  \
140           24                     24                0  CLOSED   

            last_update        date  hour          latitude         longitude  
140 2024-12-20 07:24:20  2024-12-20     7  45.7772936957677  4.87675865740856

In [411]:
data[data["name"] == "14005 - BRON - UNIVERSITE"]

number                       name               address  \
343   14005  14005 - BRON - UNIVERSITE  No Adresss Specified   

                               position  banking  bonus  bike_stands  \
343  {'lat': 45.72284, 'lng': 4.915219}     True  False           30   

     available_bike_stands  available_bikes status         last_update  \
343                      0                0   OPEN 2024-12-17 13:20:22   

           date  hour  latitude longitude  
343  2024-12-17    13  45.72284  4.915219

In [413]:
data[data["name"] == "10049 - BORNE TEST A. THOMAS"]

number                          name                   address  \
144   10049  10049 - BORNE TEST A. THOMAS  Borne test Albert Thomas   

                                              position  banking  bonus  \
144  {'lat': 45.7481858411998, 'lng': 4.86445471733...    False  False   

     bike_stands  available_bike_stands  available_bikes status  \
144            1                      0                0   OPEN   

            last_update        date  hour          latitude         longitude  
144 2024-11-22 09:50:59  2024-11-22     9  45.7481858411998  4.86445471733799

Looking at the data we can observe that the station '10124 - SALENGRO / YVONNE' is permanently closed as specified by the status column which is one of the prominent reason for it's zero bike availability. 

Let's try to find the reason for the other two in the next step:

In [416]:
data[(data["available_bike_stands"] == 0) & (data["available_bikes"] == 0)]

#Using the & operator in pandas dataframe to check 2 conditions at a time

number                          name                        address  \
144      10049  10049 - BORNE TEST A. THOMAS       Borne test Albert Thomas   
343      14005     14005 - BRON - UNIVERSITE           No Adresss Specified   
141493    9020            9020 - BOURBONNAIS  Angle rue Horand / rue Masset   

                                                 position  banking  bonus  \
144     {'lat': 45.7481858411998, 'lng': 4.86445471733...    False  False   
343                    {'lat': 45.72284, 'lng': 4.915219}     True  False   
141493                {'lat': 45.773468, 'lng': 4.800286}    False  False   

        bike_stands  available_bike_stands  available_bikes status  \
144               1                      0                0   OPEN   
343              30                      0                0   OPEN   
141493           12                      0                0   OPEN   

               last_update        date  hour          latitude  \
144    2024-11-22 09:50:59  2024-11-22     9  45.7481858411998   
343    2024-12-17 13:20:22  2024-12-17    13          45.72284   
141493 2024-12-23 10:31:29  2024-12-23    10         45.773468   

               longitude  
144     4.86445471733799  
343             4.915219  
141493          4.800286

The above dataframe represents the bike stations where the availability of bikes as well as bike stands is ZERO. The banking attribute could be a possible reason for this which specifies the credit card payment methods. We will analyze the Banking attribute in the later stage of the code.

### 6. How many stations are "CLOSED" in the dataset?

In [420]:
closed = data[data['status'] == 'CLOSED']
closed_grouped = closed.groupby(by='name')['available_bikes'].mean()

In [422]:
closed_grouped

name
10124 - SALENGRO / YVONNE           0.000000
28002 - FONTAINES S/S - GAMBETTA    7.771552
Name: available_bikes, dtype: float64

These are the bike stations in Lyon that are closed as per our dataset.

### 7. Hourly Analysis of Bike Stations with 0 Available Bikes

In [426]:
df5 = data.groupby(by=['name', 'hour'])['available_bikes'].sum().sort_values(ascending=False)
top_bike_stations5 = df5.reset_index()
zero_bike_stations = top_bike_stations5[top_bike_stations5['available_bikes'] == 0]
zero_bike_stations_sorted = zero_bike_stations.sort_values(by='hour')

In [428]:
zero_bike_stations_sorted

name  hour  available_bikes
6087     10124 - SALENGRO / YVONNE     7                0
6094     10071 - LYCÉE MARIE CURIE     9                0
6095  10049 - BORNE TEST A. THOMAS     9                0
6099        10024 - MAISONS NEUVES     9                0
6085               1024 - ROUVILLE     9                0
...                            ...   ...              ...
6141               4021 - TABAREAU    23                0
6144               3090 - VINATIER    23                0
6146   4006 - PHILIPPE DE LASSALLE    23                0
6149           4002 - MAIRIE DU 4E    23                0
6181                9052 - BALMONT    23                0

[166 rows x 3 columns]

### 8. At which hour throughout the dataset, the number of available bikes was 0?

In [431]:
df6 = data.groupby(by='hour')['available_bikes'].sum().sort_values(ascending=False)
top_bike_stations6 = df6.reset_index()
zero_hours = top_bike_stations6[top_bike_stations6['available_bikes'] == 0]
zero_hours

hour  available_bikes
16     7                0

We have not collected any data before 9 on any of the three days so we need to check the occurence of the hour 7 in the dataset to understand if it is an outlier.

In [434]:
x = top_bike_stations6[top_bike_stations6['hour'] == 7] #checking if it's an outlier
x

hour  available_bikes
16     7                0

Declaring hour 7 to be an outlier. Hence we will remove the row with hour 7 to be accurate in our analysis.

In [437]:
data = data[data['hour'] != 7 ]

In [454]:
data.head(2)

number                       name  \
0    2010  2010 - CONFLUENCE / DARSE   
1    5015           5015 - FULCHIRON   

                                         address  \
0  ANGLE ALLEE ANDRE MURE ET QUAI ANTOINE RIBOUD   
1           Devant le n°41 rue de la Quarantaine   

                              position  banking  bonus  bike_stands  \
0  {'lat': 45.743317, 'lng': 4.815747}     True  False           22   
1   {'lat': 45.75197, 'lng': 4.821662}     True  False           19   

   available_bike_stands  available_bikes status         last_update  \
0                     21                1   OPEN 2024-12-21 09:59:28   
1                      6               11   OPEN 2024-12-21 10:04:01   

         date  hour   latitude longitude  
0  2024-12-21     9  45.743317  4.815747  
1  2024-12-21    10   45.75197  4.821662

### 9. Analyzing the Hourly Trends in Bike Availability

In [492]:
df8 = data.groupby(by='hour')['available_bikes'].mean()
hourly_available = df8.reset_index().sort_values(by='hour')
hourly_available

hour  available_bikes
0      9         9.229337
1     10         9.486834
2     11         9.459293
3     12         9.522129
4     13         9.591552
5     14         9.465672
6     15         9.530971
7     16         9.487697
8     17         9.622419
9     18         9.609892
10    19         9.649125
11    20         9.673888
12    21         9.667960
13    22         9.557377
14    23         9.534687

### 10. What times of day have the highest and lowest average bike availability?

In [495]:
max_available_bikes = hourly_available['available_bikes'].max()
max_rows = hourly_available[hourly_available['available_bikes'] == max_available_bikes]
max_rows

hour  available_bikes
11    20         9.673888

08:00 PM is the time when the bike availability is the highest

In [498]:
min_available_bikes = hourly_available['available_bikes'].min()
min_rows = hourly_available[hourly_available['available_bikes'] == min_available_bikes]
min_rows

hour  available_bikes
0     9         9.229337

We have not collected any data before 9 on any of the three days so establishing hour 6 also as an outlier and dropping all the instances with hour 6.

In [501]:
data = data[data['hour'] != 6 ]

In [503]:
data.head(2)

number                       name  \
0    2010  2010 - CONFLUENCE / DARSE   
1    5015           5015 - FULCHIRON   

                                         address  \
0  ANGLE ALLEE ANDRE MURE ET QUAI ANTOINE RIBOUD   
1           Devant le n°41 rue de la Quarantaine   

                              position  banking  bonus  bike_stands  \
0  {'lat': 45.743317, 'lng': 4.815747}     True  False           22   
1   {'lat': 45.75197, 'lng': 4.821662}     True  False           19   

   available_bike_stands  available_bikes status         last_update  \
0                     21                1   OPEN 2024-12-21 09:59:28   
1                      6               11   OPEN 2024-12-21 10:04:01   

         date  hour   latitude longitude  
0  2024-12-21     9  45.743317  4.815747  
1  2024-12-21    10   45.75197  4.821662

In [505]:
min_available_bikes = hourly_available['available_bikes'].min()
min_rows = hourly_available[hourly_available['available_bikes'] == min_available_bikes]
min_rows

hour  available_bikes
0     9         9.229337

09:00 AM is the time when the bike availability is the lowest

### 11. Identifying Credit Card Payment-Only Stations

In [161]:
df_banking = data[data['banking'] == True]
unique_names = df_banking['name'].unique()
len(unique_names)

207

In [163]:
pd.DataFrame(unique_names)

0
0                      2010 - CONFLUENCE / DARSE
1                               5015 - FULCHIRON
2                        32001 - COUZON - CENTRE
3                                    6004 - FOCH
4                  7035 - MARSEILLE / UNIVERSITÉ
..                                           ...
202                         9004 - GARE DE VAISE
203                8002- PLACE AMBROISE COURTOIS
204                         9033 - GORGE DE LOUP
205  5001 - PLACE VARILLON (FUNICULAIRE ST JUST)
206                     3018 - CRÉQUI / VOLTAIRE

[207 rows x 1 columns]

Out of all the 416 bike stations in Lyon, 207 stations have credit card facility. Stations with credit card facilities are likely to have higher utilization rates, as users prefer these stations.

### 12. Identifying Stations that give incentives for returning bikes

In [166]:
df_bonus = data[data['bonus'] == True]
unique_stations = df_bonus['name'].unique()
len(unique_stations)

43

In [168]:
pd.DataFrame(unique_stations)

0
0                5016 - POINT DU JOUR / GRANGES
1                               4025 - CHAZIÈRE
2                                  4004 - HÉNON
3                1023 - CROIX ROUSSE / PERFETTI
4           4009 - JACQUARD / DENFERT-ROCHEREAU
5                          5045 - VALDO / RIVET
6                                1036 - AVEYRON
7                               4021 - TABAREAU
8                     9009 - BERLIOZ / CHALINEL
9                   4006 - PHILIPPE DE LASSALLE
10                        4042 - GREGORY COUPET
11                            5053 - SAINT-JUST
12                          4002 - MAIRIE DU 4E
13                      4007 - PLACE FLAMMARION
14                30002 - RILLIEUX - LOUP PENDU
15                   23001 - TASSIN - TUILERIES
16                               9052 - BALMONT
17              4022 - PLACE DE LA CROIX-ROUSSE
18            19001 - ST GENIS L. - MÉDIATHÈQUE
19                      24001 - ECULLY - MAIRIE
20                             5044 - CHAMPVERT
21      4014 - PHILIPPE DE LASSALLE / PILLEMENT
22    5002 - PLACE DES COMPAGNONS DE LA CHANSON
23                   5055 - BUYER / APOLLINAIRE
24               4041 - BOUSSANGES / AUSTERLITZ
25                          4008 - CLOS SAVARON
26                         4005 - PLACE BERTONE
27                     9050 - ALLEE DE L'EUROPE
28                        5007 - PLACE DE TRION
29              4003 - PLACE COMMANDANT  ARNAUD
30                       4017 - PLACE DES TAPIS
31                  00202 - PLATINE PLAISIR BEE
32                       9051 - ANDREI SAKHAROV
33                  30001 - RILLIEUX - CANELLAS
34              11001 - CALUIRE - GARE DE CUIRE
35                             5054 - CHAMPVERT
36                   4011 - PLACE JOANNÈS AMBRE
37                         5008 - SAINTE-IRÉNÉE
38                 5031- MÉNIVAL / JOLIOT CURIE
39                               25001 - ST CYR
40                            5036 - LA SALETTE
41     5030 - CHARCOT / RÉSIDENCE UNIVERSITAIRE
42  5001 - PLACE VARILLON (FUNICULAIRE ST JUST)

Out of all the 416 bike stations in Lyon, 43 stations have incentives for returning bikes. Stations with incentives are also likely to have higher utilization rates, as users prefer these stations.

### 13. How does bike usage of each station vary throughout the day? 

In [517]:
daily_availability = data.groupby(['name', 'hour'])['available_bikes'].mean().reset_index()
pivot_table = daily_availability.pivot(index='name', columns='hour', values='available_bikes')
pivot_table

hour                                           9          10         11  \
name                                                                      
0-555 - ATELIER VÉLO'V                   0.000000   0.055556   0.000000   
00202 - PLATINE PLAISIR BEE              2.000000   2.000000   2.000000   
10001 - IUT / FEYSSINE                   6.500000  11.000000   9.500000   
10002 - INSA                             8.333333  11.952381  11.631579   
10004 - TONKIN                           3.428571   4.200000   3.000000   
...                                           ...        ...        ...   
9046 - MANGINI                          21.000000  16.944444  17.750000   
9049 - SERGENT BERTHET / GORGE DE LOUP   5.500000   2.375000   2.000000   
9050 - ALLEE DE L'EUROPE                 7.000000   3.625000   3.769231   
9051 - ANDREI SAKHAROV                   6.000000   2.400000   2.769231   
9052 - BALMONT                           4.800000   2.571429   3.000000   

hour                                           12         13         14  \
name                                                                      
0-555 - ATELIER VÉLO'V                   0.000000   0.050000   0.000000   
00202 - PLATINE PLAISIR BEE              2.000000   2.000000   2.000000   
10001 - IUT / FEYSSINE                   6.400000   5.444444   6.277778   
10002 - INSA                            10.291667   8.521739   9.300000   
10004 - TONKIN                           3.222222   3.411765   2.952381   
...                                           ...        ...        ...   
9046 - MANGINI                          17.173913  17.210526  17.473684   
9049 - SERGENT BERTHET / GORGE DE LOUP   1.315789   1.666667   1.600000   
9050 - ALLEE DE L'EUROPE                 3.117647   3.200000   2.842105   
9051 - ANDREI SAKHAROV                   2.000000   2.000000   2.100000   
9052 - BALMONT                           2.000000   2.000000   2.000000   

hour                                           15         16         17  \
name                                                                      
0-555 - ATELIER VÉLO'V                   0.000000   0.000000   0.000000   
00202 - PLATINE PLAISIR BEE              2.000000   2.000000   2.000000   
10001 - IUT / FEYSSINE                   6.761905   3.200000   2.222222   
10002 - INSA                             8.913043   8.521739   7.173913   
10004 - TONKIN                           1.950000   2.900000   1.684211   
...                                           ...        ...        ...   
9046 - MANGINI                          17.000000  16.000000  15.684211   
9049 - SERGENT BERTHET / GORGE DE LOUP   1.052632   1.136364   1.227273   
9050 - ALLEE DE L'EUROPE                 2.842105   2.666667   2.842105   
9051 - ANDREI SAKHAROV                   2.000000   2.000000   2.210526   
9052 - BALMONT                           2.000000   1.800000   1.666667   

hour                                           18         19         20  \
name                                                                      
0-555 - ATELIER VÉLO'V                   0.000000   0.000000   0.000000   
00202 - PLATINE PLAISIR BEE              2.000000   2.000000   2.000000   
10001 - IUT / FEYSSINE                   2.631579   3.157895   2.777778   
10002 - INSA                             7.526316   7.826087  10.000000   
10004 - TONKIN                           2.454545   2.904762   3.333333   
...                                           ...        ...        ...   
9046 - MANGINI                          16.173913  16.863636  17.000000   
9049 - SERGENT BERTHET / GORGE DE LOUP   1.470588   2.380952   2.529412   
9050 - ALLEE DE L'EUROPE                 2.666667   2.666667   2.666667   
9051 - ANDREI SAKHAROV                   2.300000   2.333333   2.166667   
9052 - BALMONT                           1.523810   1.210526   0.947368   

hour                                           21    22    23  
name                            

### 14. How does bike usage of each station varied across the three days in our dataset?

In [526]:
daily_availability = data.groupby(['name', 'date'])['available_bikes'].mean().reset_index()
pivot_table = daily_availability.pivot(index='name', columns='date', values='available_bikes')
pivot_table

date                                    2024-11-22  2024-12-17  2024-12-21  \
name                                                                         
0-555 - ATELIER VÉLO'V                         NaN         NaN    0.000000   
00202 - PLATINE PLAISIR BEE                    NaN         NaN    2.000000   
10001 - IUT / FEYSSINE                         NaN         NaN    7.819277   
10002 - INSA                                   NaN         NaN   10.899083   
10004 - TONKIN                                 NaN         NaN    4.845238   
...                                            ...         ...         ...   
9046 - MANGINI                                 NaN         NaN   14.364706   
9049 - SERGENT BERTHET / GORGE DE LOUP         NaN         NaN    0.779221   
9050 - ALLEE DE L'EUROPE                       NaN         NaN    1.000000   
9051 - ANDREI SAKHAROV                         NaN         NaN    0.180556   
9052 - BALMONT                                 NaN         NaN    0.000000   

date                                    2024-12-22  2024-12-23  
name                                                            
0-555 - ATELIER VÉLO'V                    0.000000    0.024390  
00202 - PLATINE PLAISIR BEE               2.000000    2.000000  
10001 - IUT / FEYSSINE                    3.440000    3.378378  
10002 - INSA                             11.897727    4.769231  
10004 - TONKIN                            2.311688    1.487179  
...                                            ...         ...  
9046 - MANGINI                           17.000000   19.528736  
9049 - SERGENT BERTHET / GORGE DE LOUP    2.303797    2.753247  
9050 - ALLEE DE L'EUROPE                  1.000000    6.333333  
9051 - ANDREI SAKHAROV                    0.000000    6.000000  
9052 - BALMONT                            0.000000    5.077922  

[414 rows x 5 columns]

As we have only tasken

In [161]:
data['bike_stands'].unique()
bins = [0, 20, 40 , 60]
groups = ["Small Station", "Medium Station", "Big Station"]
binned = pd.cut(data["bike_stands"], bins, labels = groups)

In [163]:
binned.unique()

['Medium Station', 'Small Station', 'Big Station']
Categories (3, object): ['Small Station' < 'Medium Station' < 'Big Station']

In [181]:
Station_type = pd.DataFrame({"Station Names": data["name"], "Bike Stands": data["bike_stands"], "Station Type": binned})

grouped_station = Station_type.groupby("Station Names").agg(bike_stands=("Bike Stands", "first"), station_type=("Station Type", "first")).reset_index()
grouped_station

Station Names  bike_stands    station_type
0                    0-555 - ATELIER VÉLO'V            3   Small Station
1               00202 - PLATINE PLAISIR BEE            6   Small Station
2                    10001 - IUT / FEYSSINE           30  Medium Station
3                              10002 - INSA           55     Big Station
4                            10004 - TONKIN           10   Small Station
..                                      ...          ...             ...
411                          9046 - MANGINI           24  Medium Station
412  9049 - SERGENT BERTHET / GORGE DE LOUP           16   Small Station
413                9050 - ALLEE DE L'EUROPE           12   Small Station
414                  9051 - ANDREI SAKHAROV           12   Small Station
415                          9052 - BALMONT           12   Small Station

[416 rows x 3 columns]

In [331]:
data.set_index(['name', 'date']).sort_values('name', ascending = False)

number               address  \
name                   date                                       
9052 - BALMONT         2024-12-22    9052    360 BVD DE BALMONT   
                       2024-12-22    9052    360 BVD DE BALMONT   
                       2024-12-23    9052    360 BVD DE BALMONT   
                       2024-12-21    9052    360 BVD DE BALMONT   
                       2024-12-21    9052    360 BVD DE BALMONT   
...                                   ...                   ...   
0-555 - ATELIER VÉLO'V 2024-12-23     555  No Adresss Specified   
                       2024-12-22     555  No Adresss Specified   
                       2024-12-21     555  No Adresss Specified   
                       2024-12-23     555  No Adresss Specified   
                       2024-12-23     555  No Adresss Specified   

                                                              position  \
name                   date                                              
9052 - BALMONT         2024-12-22  {'lat': 45.789194, 'lng': 4.799308}   
                       2024-12-22  {'lat': 45.789194, 'lng': 4.799308}   
                       2024-12-23  {'lat': 45.789194, 'lng': 4.799308}   
                       2024-12-21  {'lat': 45.789194, 'lng': 4.799308}   
                       2024-12-21  {'lat': 45.789194, 'lng': 4.799308}   
...                                                                ...   
0-555 - ATELIER VÉLO'V 2024-12-23  {'lat': 45.780403, 'lng': 4.904869}   
                       2024-12-22  {'lat': 45.780403, 'lng': 4.904869}   
                       2024-12-21  {'lat': 45.780403, 'lng': 4.904869}   
                       2024-12-23  {'lat': 45.780403, 'lng': 4.904869}   
                       2024-12-23  {'lat': 45.780403, 'lng': 4.904869}   

                                   banking  bonus  bike_stands  \
name                   date                                      
9052 - BALMONT         2024-12-22    False   True           12   
                       2024-12-22    False   True           12   
                       2024-12-23    False   True           12   
                       2024-12-21    False   True           12   
                       2024-12-21    False   True           12   
...                                    ...    ...          ...   
0-555 - ATELIER VÉLO'V 2024-12-23    False  False            3   
                       2024-12-22    False  False            3   
                       2024-12-21    False  False            3   
                       2024-12-23    False  False            3   
                       2024-12-23    False  False            3   

                                   available_bike_stands  available_bikes  \
name                   date                                                 
9052 - BALMONT         2024-12-22                     12                0   
                       2024-12-22                     12                0   
                       2024-12-23                      7                5   
                       2024-12-21                     12                0   
                       2024-12-21                     12                0   
...                                                  ...              ...   
0-555 - ATELIER VÉLO'V 2024-12-23                      3                0   
                       2024-12-22                      3                0   
                       2024-12-21                      3                0   
                       2024-12-23                      3                0   
                       2024-12-23                      3                0   

                                  status         last_update  hour   latitude  \
name                   date                                                     
9052 - BALMONT         2024-12-22   OPEN 2024-12-22 18:39:29    18  45.789194   
                       2024-12-22   OPEN 2024-12-22 15:35:27    15  45.789194   
               

In [335]:
# Filtering the data to identify rows where 'available_bike_stands' is less than 1
bike_shortage = data[data.available_bike_stands < 1]
bike_shortage_groupped = bike_shortage.groupby('name').count().sort_values(['available_bike_stands'], ascending=False)['available_bike_stands']
print(bike_shortage_groupped)

name
00202 - PLATINE PLAISIR BEE            239
20003 - OULLINS - PONT BLANC           157
2025 - MERCIÈRE / FERRANDIÈRE          151
2003 - PLACE AMPÈRE                    141
3038 - VIVIER MERLE / PAUL BERT EST    137
                                      ... 
10049 - BORNE TEST A. THOMAS             1
15002 - ST PRIEST - CITE BERLIET         1
9020 - BOURBONNAIS                       1
14005 - BRON - UNIVERSITE                1
3100 - SISLEY / DAUPHINE                 1
Name: available_bike_stands, Length: 155, dtype: int64


In [ ]:
Analysis

The body of the notebook should consist of the analysis work to carry out the plan and 
be presented in the style outlined above.

Conclusion
At the end of the notebook there should be a conclusion, discussing how the results of 
the analysis compare with the goals set out in the plan (3 paragraphs of text).

Percentage Contribution:
Sumit Trivedi   - 
Moutrisha Ghosh - 